[FLOC-1373] Switch to PEP 440 compliant versioning.

tomprince opened the pull request:
https://clusterhq.atlassian.net/browse/FLOC-1373

I changed the method of expressing doc releases. There can only be one `+` in a PEP440 version number. Thus, if a tag already has a `+` in it, versioneer will use a `.` instead of a `+` to indicate the git revision. Rather than try to deal with this, I switched to using [post](https://www.python.org/dev/peps/pep-0440/#post-releases) releases, which capture the semantics of a doc release.

The versioneer changes can be verified by installing the new version of versioneer (specified in setup.py), running `versioneer install`, and verifying that it is idempotent.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/clusterhq/flocker/1699)
<!-- Reviewable:end -->

adamtheturtle commented on setup.py:
```diff
@@ -83,7 +71,7 @@
     # This is necessary for a release because our version scheme does not
```
Remove this comment

adamtheturtle commented on versioneer.py:
```diff
@@ -1,5 +1,5 @@
 
-# Version: 0.10
+# Version: 0.15
```
Please confirm that these changes are created by the tool and so I can ignore them in the review.

tomprince replied:
I described above one way to test this.

adamtheturtle commented:
See:

```python
def upload_python_packages(scratch_directory, target_bucket, top_level,
                           output, error):
    """
...
    """
    if setuptools_version != '3.6':
        # XXX Use PEP440 version system so new setuptools can be used.
        # https://clusterhq.atlassian.net/browse/FLOC-1331.
        raise IncorrectSetuptoolsVersion()
```

adamtheturtle commented on admin/test/test_packaging.py:
```diff
@@ -542,19 +542,16 @@ def test_version_found(self):
-            '0.3.2+doc1',
-            '0.3.2-1-gf661a6a',
-            '0.3.2+doc1-1-gf661a6a',
+            '0.3.2+post1',
```
`0.3.2+post1` and alike are not valid versions - replace `+post` with `.post` wherever it exists.

adamtheturtle commented:
`packaging.py` includes:

```python
        # We require an old version of setuptools
        # XXX This should not be necessary after
        # https://clusterhq.atlassian.net/browse/FLOC-1373
        'backup-file-in-package /opt/flocker/lib/python2.7/site-packages/setuptools-3.6.dist-info/requires.txt.orig',  # noqa
```

adamtheturtle commented:
I assume that the Buildbot changes mean that we won't be able do do any docs releases for currently released versions (up to and including 1.0.1). I assume that this is fine but would like it confirmed with the team.

adamtheturtle commented on setup.py:
```diff
@@ -47,11 +35,11 @@
     # Some of the tests use Conch:
     "PyCrypto==2.6.1",
-    "pyasn1==0.1.7",
+    "pyasn1",
```
Why does this no longer have a pinned version? Either pin a version or add a comment.

adamtheturtle commented on flocker/node/functional/sendbytes-docker/Dockerfile.in:
```diff
@@ -1,4 +1,6 @@
 MAINTAINER ClusterHQ <support@clusterhq.com>
 ADD . /
+# If installed via wheel, the installed file will not have +x set.
+RUN ["chmod", "+x", "/run.sh"]
```
I don't understand why this change is necessary.

tomprince replied:
New versions of pip install by building a wheel. Since zip files (which is what underlies wheels) don't record `+x`, when the package gets installed, these scripts get installed without `+x`. So we need to set it here.

adamtheturtle commented on flocker/node/functional/retry-docker/Dockerfile:
```diff
@@ -2,6 +2,10 @@
-ADD run /usr/local/bin/run
 WORKDIR /data
+
+ADD run /usr/local/bin/run
```
I don't understand why this change is necessary.

itamarst replied:
What Adam said.

itamarst replied:
Ah.

jongiddy replied:
For the record, this line was not added, just moved below the WORKDIR command.  The actual change is explained in the subsequent comment.

adamtheturtle commented:
Thanks @tomprince please address the comments and resubmit.

adamtheturtle commented on flocker/common/version.py:
```diff
@@ -15,11 +15,11 @@
+    r"(\.post(?P<documentation_revision>[0-9]+))?"
     # Development version
-    r"(-(?P<commit_count>[0-9]+)-g(?P<commit_hash>[0-9a-f]+))?"
+    r"(\+(?P<commit_count>[0-9]+).g(?P<commit_hash>[0-9a-f]+))?"
```
Why are the changes to development versions / whether the tree is dirty necessary?

tomprince replied:
This is exactly the point of this change. The old versioning scheme of versioneer isn't PEP 440 compatible. The new one is.

adamtheturtle commented on flocker/common/version.py:
```diff
@@ -15,11 +15,11 @@
     # Development version
-    r"(-(?P<commit_count>[0-9]+)-g(?P<commit_hash>[0-9a-f]+))?"
+    r"(\+(?P<commit_count>[0-9]+).g(?P<commit_hash>[0-9a-f]+))?"
     # Wether the tree is dirty.
```
:question: existing typo: Wether > Whether.

tomprince commented:
> I assume that the Buildbot changes mean that we won't be able do do any docs releases for currently released versions (up to and including 1.0.1). I assume that this is fine but would like it confirmed with the team.

As I mentioned in that PR, I specifically left support for the old versioning scheme as well in buildbot, for exactly this purpose.

adamtheturtle commented:
:question: Put a comment above `_VERSION_RE` describing the PEP 440 limits which should be accounted for in any new version type.

adamtheturtle commented on admin/release.py:
```diff
@@ -658,11 +650,6 @@ def upload_python_packages(scratch_directory, target_bucket, top_level,
     :param bytes target_bucket: S3 bucket to upload packages to.
     :param FilePath top_level: The top-level of the flocker repository.
     """
-    if setuptools_version != '3.6':
```
:question: Is it not useful to have the equivalent of this, but for the new necessary setuptools version?

tomprince replied:
No. We had a fixed version of setuptools because there was a specific change in behavior that we needed avoid. We should be able to work with any recent version of setuptools, now.

adamtheturtle commented on flocker/common/test/test_version.py:
```diff
@@ -146,269 +80,207 @@ def test_invalid_Version(self):
-
-
-class IsPreReleaseTests(SynchronousTestCase):
+def build_version_test(name, version_case):
```
:smile:

adamtheturtle commented on flocker/common/test/test_version.py:
```diff
@@ -146,269 +80,207 @@ def test_invalid_Version(self):
-        When the documentation version is from a documentation release, it is a
+    :param FlockerVersion flocker_version: The parsed version.
+    :param bytes doc_version: The corresponding doc version.
+    :param bytes installable_version: The corresponding doc version.
```
Incorrect description

adamtheturtle commented on flocker/common/test/test_version.py:
```diff
@@ -146,269 +80,207 @@ def test_invalid_Version(self):
+            commit_hash='f661a6a',
+            dirty='.dirty',
+        ),
+        doc_version="0.3.2.post11+1.gf661a6a.dirty",
```
These are explicitly bytes in the `build_version_test ` docstring, should they be explicitly bytes here?

adamtheturtle commented on docs/gettinginvolved/infrastructure/release-schedule-versioning.rst:
```diff
@@ -77,7 +77,7 @@ Documentation releases will be made when documentation for a major or minor mark
 Documentation releases will be planned and scheduled by ClusterHQ's product team, in consultation with the marketing and engineering teams.
 
-The version of a documentation will have the version of the corresponding marketing release, with a ``+docX`` release, where ``X`` starts at ``1`` and is incremented for each documentation release.
+The version of a documentation will have the version of the corresponding marketing release, with a ``.postX`` release, where ``X`` starts at ``1`` and is incremented for each documentation release.
```
:question: Add a legacy section mentioning the old doc version style.

tomprince replied:
I think if somebody needs to look at historical versions, they can look at the documentation from a corresponding release.

adamtheturtle commented:
Thanks @tomprince , address the new comments and merge when:
* `master` is open for merges
* The necessary BuildBot changes are ready
* The build is green (enough)

wallrj commented:
I'm so looking forward to not having to type `pip install setuptools==3.6`. This'll fix that, right?

tomprince commented:
I've update this to use PEP440 normalized versions everywhere. (Setuptools creates sdists with normalized versions, and installs dist-info with normalized versions, so I chose not to fight to figure out where we need normalized versions and where not.

tomprince commented:
The homebrew recipe creation failure is #1748.

itamarst commented:
Adam seemed happy with this in previous iteration, given addressed comments. Any reason not to proceed with merging, @tomprince ?

tomprince commented:
> Adam seemed happy with this in previous iteration, given addressed comments. Any reason not to proceed with merging, @tomprince ?

I ended up needing to make other changes to the versioning scheme, and wanted another pair of eyes on the changes.

itamarst commented on flocker/common/test/test_version.py:
```diff
@@ -146,269 +80,271 @@ def test_invalid_Version(self):
+    is interpreted as expected.
     """
+    class Tests(SynchronousTestCase):
+        def test_flocker_version(self):
```
Can you add quick docstrings for these tests?

jongiddy replied:
Added by @tomprince

itamarst commented:
Urgh. I'll finish this up tomorrow, not awake enough to parse regexs in my head.

zubron commented:
There is one failing test and it seem to be relevant to this change.

```
[ERROR]
Traceback (most recent call last):
  File "/opt/flocker/lib/python2.7/site-packages/twisted/trial/runner.py", line 497, in loadPackage
    module = modinfo.load()
  File "/opt/flocker/lib/python2.7/site-packages/twisted/python/modules.py", line 383, in load
    return self.pathEntry.pythonPath.moduleLoader(self.name)
  File "/opt/flocker/lib/python2.7/site-packages/twisted/python/reflect.py", line 303, in namedAny
    topLevelPackage = _importAndCheckStack(trialname)
  File "/opt/flocker/lib/python2.7/site-packages/twisted/python/reflect.py", line 250, in _importAndCheckStack
    reraise(excValue, excTraceback)
  File "/opt/flocker/lib/python2.7/site-packages/flocker/common/test/test_version.py", line 10, in <module>
    from packaging.version import Version as PEP440Version
exceptions.ImportError: No module named packaging.version

flocker.common.test.test_version
```

zubron commented on admin/test/test_packaging.py:
```diff
@@ -542,19 +542,16 @@ def test_version_found(self):
         """
         versions = [
             '0.3.2',
             '0.3.3dev5',
```
There is a mix of "dev5" and ".dev5" being used. Is this intentional and if so, can you explain why?

zubron commented on admin/test/test_release.py:
```diff
@@ -697,7 +694,7 @@ def test_production_can_publish_prerelease(self):
         # Does not raise:
         self.publish_docs(
-            aws, '0.3.2pre1', '0.3.2pre1', environment=Environments.PRODUCTION)
+            aws, '0.3.2rc1', '0.3.2rc1', environment=Environments.PRODUCTION)
```
Only some instances of "pre" have been changed to "rc".

zubron replied:
```
$ grep -rHn --include "*.py" [0-9]pre .
./common/test/test_version.py:311:        version=b'0.3.2pre11+1.gf661a6a',
./common/test/test_version.py:320:        doc_version=b'0.3.2pre11+1.gf661a6a',
./common/test/test_version.py:321:        installable_version=b'0.3.2pre11',
```

jongiddy replied:
The remaining tests containing `pre` check for correct behaviour with legacy version numbers.

adamtheturtle commented on flocker/common/version.py:
```diff
@@ -250,7 +253,7 @@ def make_rpm_version(flocker_version):
     # 0.1.2pre2  becomes
```
Outdated comment.

adamtheturtle commented on flocker/common/version.py:
```diff
@@ -11,15 +11,18 @@
     # The base version
     r"(?P<major>[0-9])\.(?P<minor>[0-9]+)\.(?P<micro>[0-9]+)"
     # Pre-release
```
Comment that this also supports a legacy system and why.

adamtheturtle commented on flocker/common/version.py:
```diff
@@ -11,15 +11,18 @@
     # The base version
```
Comment that these is a PEP440 compatible system.

jongiddy replied:
Comment added by @tomprince

tomprince commented:
Review status: 0 of 19 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

<sup>**[admin/test/test_release.py, line 697 \[r8\]](https://reviewable.io:443/reviews/clusterhq/flocker/1699#-JtjSVtAfujdGSgDuJ14-r8-697)** ([raw file](https://github.com/clusterhq/flocker/blob/4b768ac8437a1c913aef5fc7be88a48b0b281702/admin/test/test_release.py#L697)):</sup>
The ones quoted are explictly tests that we can still parse the old versions. (This is necessary for as long as old tags are the most recent tags on open branches). 

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/clusterhq/flocker/1699)
<!-- Sent from Reviewable.io -->

tomprince commented:
I won't be able to address any review feedback.

jongiddy commented:
All comments have been addressed. I have added comments here to confirm some that need no further code changes.

I have made a simple change to support Flocker 10 and above, and set Versioneer to 0.15 in `dev-requirements.txt` so generated files do not get changed if `versioneer install` is re-run.

wallrj commented:
I suppose in hindsight it would have been better to upgrade versioneer in a separate step.

My eyes glossed over a bit towards the end, but I think it's OK to merge this when the tutorial box has built and the triggered acceptance tests have passed.

tomprince commented:
> I suppose in hindsight it would have been better to upgrade versioneer in a separate step.

I'm not sure that would be entirely possible. `versioneer` changed how non-tagged commits are versioned, and much of the changes here are addressing that.
